ReplayTests: Add support for >=gtest-1.10

ConiKost opened the pull request:
I am aware, that you don't recommend to use systems gtest, but it seems, there has been some changes.
Up to gtest-1.9.0, all tests are running fine.

With gtest-1.10 and newer, the ReplayTests will fail. Setting, as suggested, makes the test running fine. I don't know, if there is a proper fix for the tests. 

```
12/14 Testing: replay_tests
12/14 Test: replay_tests
Command: "/var/tmp/portage/games-simulation/openrct2-0.2.6/work/openrct2-0.2.6_build/test_replays"
Directory: /var/tmp/portage/games-simulation/openrct2-0.2.6/work/openrct2-0.2.6_build
"replay_tests" start time: Jul 21 03:28 CEST
Output:
----------------------------------------------------------
Running main() from /var/tmp/portage/dev-cpp/gtest-1.10.0_p20200702/work/googletest-aee0f9d9b5b87796ee8a0ab26b7587ec30e8858e/googletest/src/gtest_main.cc
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from GoogleTestVerification
[ RUN      ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<ReplayTests>
/var/tmp/portage/games-simulation/openrct2-0.2.6/work/OpenRCT2-0.2.6/test/tests/ReplayTests.cpp:72: Failure
Parameterized test suite ReplayTests is defined via TEST_P, but never instantiated. None of the test cases will run. Either no INSTANTIATE_TEST_SUITE_P is provided or the only ones provided expand to nothing.

Ideally, TEST_P definitions should only ever be included as part of binaries that intend to use them. (As opposed to, for example, being placed in a library that may be linked in to get other utilities.)

To suppress this error for this test suite, insert the following line (in a non-header) in the namespace it is defined in:

GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(ReplayTests);
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<ReplayTests> (0 ms)
[----------] 1 test from GoogleTestVerification (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] GoogleTestVerification.UninstantiatedParameterizedTestSuite<ReplayTests>

 1 FAILED TEST
<end of output>
Test time =   0.01 sec
----------------------------------------------------------
Test Failed.
"replay_tests" end time: Jul 21 03:28 CEST
"replay_tests" time elapsed: 00:00:00
----------------------------------------------------------
```

Signed-off-by: Conrad Kostecki <conrad@kostecki.com>

ConiKost commented:
Well, that seems not to work with the included gtest. Is there any possiblity to make that version depended?

janisozaur commented:
I tried working around an API change in one of the older versions in the past and couldn't find any version define, so I suppose you'd have to consume the version from CMake.

While at it, you can also remove the warning and maybe switch the option on gtest too - I found it a bit ungrounded and it seems gtest maintainers eventually did so as well and removed it - try following the link now.

janisozaur commented:
> couldn't find any version define

That was in the old version though, things may have changed or I simply have not looked hard enough.

ConiKost commented:
> I tried working around an API change in one of the older versions in the past and couldn't find any version define, so I suppose you'd have to consume the version from CMake.

Despite of my change: Maybe it's a good idea to update the shipped gtest?

> While at it, you can also remove the warning and maybe switch the option on gtest too - I found it a bit ungrounded and it seems gtest maintainers eventually did so as well and removed it - try following the link now.

Eh? I am sorry, but I don't understand. You mean the switch to use system installed gtest? Which link do you mean?

janisozaur commented:
This link: https://github.com/OpenRCT2/OpenRCT2/blob/22f29bf192320ef78cabd8bc76672f1e85cad038/test/tests/CMakeLists.txt#L10

janisozaur commented:
@ConiKost would you update the gtest in use? Note #12667 brings updates to CI images

ConiKost commented:
@janisozaur Could you tell me, what your CI is doing different? With the change to gtest-1.10, the build libgtest.a is now in a subdirectory `lib`. Locally, I can compile fine and all tests pass with included gtest-1.10

But the CI here has trouble:
ninja: error: 'googletest-prefix/src/googletest-build/lib/libgtest.a', needed by 'test_s6importexporttests', missing and no known rule to make it
-> ttps://github.com/ConiKost/OpenRCT2/runs/1069294682

When I drop `lib`, CI will still fail, but with not finding it.
-> ttps://github.com/ConiKost/OpenRCT2/runs/1069294682

What could I miss here, since locally the new included gtest works?!

janisozaur commented:
You haven't modified the external project

ConiKost commented:
> You haven't modified the external project

Eh, I am sorry, but I don't understand. Could you please explain further?

janisozaur commented:
https://github.com/ConiKost/OpenRCT2/blob/35019d8ccbb5fb5341c274d4d8ba63bc3926b133/test/tests/CMakeLists.txt#L31

ConiKost commented:
> https://github.com/ConiKost/OpenRCT2/blob/35019d8ccbb5fb5341c274d4d8ba63bc3926b133/test/tests/CMakeLists.txt#L31

ConiKost commented:
> https://github.com/ConiKost/OpenRCT2/blob/35019d8ccbb5fb5341c274d4d8ba63bc3926b133/test/tests/CMakeLists.txt#L31

I am still confused. I did changed the used gtest, as seen here https://github.com/OpenRCT2/OpenRCT2/pull/12735/files?
So I don't see, what you are pointing up to. I also can clearly see, that the CI seems to use the new gtest, as the path is now different and the build libs are in a lib subfolder.

janisozaur commented:
You already stated the problem:
> the path is now different and the build libs are in a lib subfolder

but you haven't updated the build information.

ConiKost commented:
> but you haven't updated the build information.

What? I am sorry, but did you read my first comment about that issue?
Locally, running the entire build process works and the files are in the lib subdirectory. So my changes does work, since I adjusted the path.

But the CI system here doesn't seem to be the case, so there is something different.
Since I can't look into the file system of the CI, I can't see, what is happening to the gtest build process. 
Is this somehow possible? This would help to understand, what is different there.

janisozaur commented:
Sure, you can look into how CI is set up: https://github.com/OpenRCT2/OpenRCT2/blob/develop/.github/workflows/ci.yml

You can follow the steps yourself or try using https://github.com/nektos/act (only learned about it recently and haven't used, so no idea if that would work)

But all that is unnecessary, you simply need to change the external project for gtest as you described already.

ConiKost commented:
> But all that is unnecessary, you simply need to change the external project for gtest as you described already.

Thanks. It seems, that I don't understand, what you exactly mean by that.
Sorry, I am just kind a bit of lost here. FWICS, we have (in git master) current `googletest`, which uses `googletest-distribution`, to build gtest. So if you don't mean by updating that, what do you mean? Drop all stuff and use systems installed gtest?

janisozaur commented:
I got a bit annoyed with the discussion and fixed it myself.

ConiKost commented:
> I got a bit impatient with the discussion and fixed it myself.

m( Sorry for wasting your time. Now I get it, what you where up to.

ConiKost commented:
Looks good, CI and tests did now passed on amd64. I've updated to a newer snapshot version, since distributions start to use newer snapshots than older 1.10.0 stable version of gtest, which also introduced more changes.

duncanspumpkin commented:
Replay tests on Linux x64 passed in 5 seconds that sounds suspicious. It normally takes about 9 seconds. The replay tests are some of the most critical tests.

ConiKost commented:
Are you sure? I can't reproduce your 9 seconds in any way. It does not matter, if I take stable, current git master or git master with my patches. It's always about ~5-6 seconds. So my changes are IMHO not related, if there is something wrong.

Stable 0.3.0:
```
Running tests...
Test project /home/conikost/build-stable
      Start  1: sawyercoding
 1/15 Test  #1: sawyercoding .....................   Passed    0.00 sec
      Start  2: languagepack
 2/15 Test  #2: languagepack .....................   Passed    0.00 sec
      Start  3: ini
 3/15 Test  #3: ini ..............................   Passed    0.00 sec
      Start  4: platform
 4/15 Test  #4: platform .........................   Passed    0.01 sec
      Start  5: string
 5/15 Test  #5: string ...........................   Passed    0.00 sec
      Start  6: localisation
 6/15 Test  #6: localisation .....................   Passed    0.00 sec
      Start  7: Crypt
 7/15 Test  #7: Crypt ............................   Passed    0.09 sec
      Start  8: ImageImporter
 8/15 Test  #8: ImageImporter ....................   Passed    0.01 sec
      Start  9: ride_ratings
 9/15 Test  #9: ride_ratings .....................   Passed    0.28 sec
      Start 10: multilaunch
10/15 Test #10: multilaunch ......................   Passed    0.82 sec
      Start 11: tile_elements
11/15 Test #11: tile_elements ....................   Passed    0.17 sec
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    0.01 sec
      Start 13: play_tests
13/15 Test #13: play_tests .......................   Passed    0.47 sec
      Start 14: pathfinding
14/15 Test #14: pathfinding ......................   Passed    0.41 sec
      Start 15: s6importexporttests
15/15 Test #15: s6importexporttests ..............   Passed    3.09 sec

100% tests passed, 0 tests failed out of 15

Total Test time (real) =   5.37 sec
```

Current Git master (without my changes):
```
Running tests...
Test project /home/conikost/build-git
      Start  1: sawyercoding
 1/15 Test  #1: sawyercoding .....................   Passed    0.00 sec
      Start  2: languagepack
 2/15 Test  #2: languagepack .....................   Passed    0.00 sec
      Start  3: ini
 3/15 Test  #3: ini ..............................   Passed    0.00 sec
      Start  4: platform
 4/15 Test  #4: platform .........................   Passed    0.01 sec
      Start  5: string
 5/15 Test  #5: string ...........................   Passed    0.00 sec
      Start  6: localisation
 6/15 Test  #6: localisation .....................   Passed    0.00 sec
      Start  7: Crypt
 7/15 Test  #7: Crypt ............................   Passed    0.09 sec
      Start  8: ImageImporter
 8/15 Test  #8: ImageImporter ....................   Passed    0.01 sec
      Start  9: ride_ratings
 9/15 Test  #9: ride_ratings .....................   Passed    0.32 sec
      Start 10: multilaunch
10/15 Test #10: multilaunch ......................   Passed    0.84 sec
      Start 11: tile_elements
11/15 Test #11: tile_elements ....................   Passed    0.18 sec
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    0.01 sec
      Start 13: play_tests
13/15 Test #13: play_tests .......................   Passed    0.49 sec
      Start 14: pathfinding
14/15 Test #14: pathfinding ......................   Passed    0.43 sec
      Start 15: s6importexporttests
15/15 Test #15: s6importexporttests ..............   Passed    2.96 sec

100% tests passed, 0 tests failed out of 15

Total Test time (real) =   5.36 sec
```

Current Git master (with my changes):
```Running tests...
Test project /home/kostecki/build-git-conikost
      Start  1: sawyercoding
 1/15 Test  #1: sawyercoding .....................   Passed    0.00 sec
      Start  2: languagepack
 2/15 Test  #2: languagepack .....................   Passed    0.00 sec
      Start  3: ini
 3/15 Test  #3: ini ..............................   Passed    0.00 sec
      Start  4: platform
 4/15 Test  #4: platform .........................   Passed    0.01 sec
      Start  5: string
 5/15 Test  #5: string ...........................   Passed    0.00 sec
      Start  6: localisation
 6/15 Test  #6: localisation .....................   Passed    0.00 sec
      Start  7: Crypt
 7/15 Test  #7: Crypt ............................   Passed    0.15 sec
      Start  8: ImageImporter
 8/15 Test  #8: ImageImporter ....................   Passed    0.01 sec
      Start  9: ride_ratings
 9/15 Test  #9: ride_ratings .....................   Passed    0.32 sec
      Start 10: multilaunch
10/15 Test #10: multilaunch ......................   Passed    0.85 sec
      Start 11: tile_elements
11/15 Test #11: tile_elements ....................   Passed    0.18 sec
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    0.01 sec
      Start 13: play_tests
13/15 Test #13: play_tests .......................   Passed    0.49 sec
      Start 14: pathfinding
14/15 Test #14: pathfinding ......................   Passed    0.43 sec
      Start 15: s6importexporttests
15/15 Test #15: s6importexporttests ..............   Passed    2.96 sec

100% tests passed, 0 tests failed out of 15

Total Test time (real) =   5.47 sec
```

janisozaur commented:
The timings are a symptom, we are interested in underlying cause. It's ok for them to be faster, but we want to ensure all the parks get tested. Can you use verbose mode of ctest to compare the tests it discovers and runs?

ConiKost commented:
Well, seems you are right. This even happens with current stable. The testsuite seems to run no tests. Can't say, why.

```
12: Test command: /home/conikost/build-0.3.0/test_replays
12: Test timeout computed to be: 10000000
12: Running main() from gtest_main.cc
12: [==========] Running 0 tests from 0 test cases.
12: [==========] 0 tests from 0 test cases ran. (0 ms total)
12: [  PASSED  ] 0 tests.
12/15 Test #12: replay_tests .....................   Passed    0.01 sec
```

janisozaur commented:
Have you ran the install target, as the docs say? Alternatively symlinked the data+testdata dirs?

ConiKost commented:
> Alternatively symlinked the data+testdata dirs?

Yes, otherwise many tests are segfaulting or failing. But I see, that testdata is symlinked already by the build system?
When I move manually the ```replays``` folder from ```OpenRCT-0.3.0/testdata/replays``` to ```OpenRCT-0.3.0/test/tests/testdata/replays there, all tests are run.

But the CI here seems to be fine, as it runs about the same?
```
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    6.23 sec
```

```
Test project /home/conikost/build-stable
      Start  1: sawyercoding
 1/15 Test  #1: sawyercoding .....................   Passed    0.00 sec
      Start  2: languagepack
 2/15 Test  #2: languagepack .....................   Passed    0.00 sec
      Start  3: ini
 3/15 Test  #3: ini ..............................   Passed    0.00 sec
      Start  4: platform
 4/15 Test  #4: platform .........................   Passed    0.01 sec
      Start  5: string
 5/15 Test  #5: string ...........................   Passed    0.00 sec
      Start  6: localisation
 6/15 Test  #6: localisation .....................   Passed    0.00 sec
      Start  7: Crypt
 7/15 Test  #7: Crypt ............................   Passed    0.22 sec
      Start  8: ImageImporter
 8/15 Test  #8: ImageImporter ....................   Passed    0.01 sec
      Start  9: ride_ratings
 9/15 Test  #9: ride_ratings .....................   Passed    0.26 sec
      Start 10: multilaunch
10/15 Test #10: multilaunch ......................   Passed    0.83 sec
      Start 11: tile_elements
11/15 Test #11: tile_elements ....................   Passed    0.17 sec
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    8.37 sec
      Start 13: play_tests
13/15 Test #13: play_tests .......................   Passed    0.47 sec
      Start 14: pathfinding
14/15 Test #14: pathfinding ......................   Passed    0.41 sec
      Start 15: s6importexporttests
15/15 Test #15: s6importexporttests ..............   Passed    2.92 sec

100% tests passed, 0 tests failed out of 15

Total Test time (real) =  13.72 sec
```

ConiKost commented:
Git Master with my PR also looks good after moving replays to the correct path.

```
Test project /home/kostecki/build-git-conikost
      Start  1: sawyercoding
 1/15 Test  #1: sawyercoding .....................   Passed    0.00 sec
      Start  2: languagepack
 2/15 Test  #2: languagepack .....................   Passed    0.00 sec
      Start  3: ini
 3/15 Test  #3: ini ..............................   Passed    0.00 sec
      Start  4: platform
 4/15 Test  #4: platform .........................   Passed    0.01 sec
      Start  5: string
 5/15 Test  #5: string ...........................   Passed    0.00 sec
      Start  6: localisation
 6/15 Test  #6: localisation .....................   Passed    0.00 sec
      Start  7: Crypt
 7/15 Test  #7: Crypt ............................   Passed    0.15 sec
      Start  8: ImageImporter
 8/15 Test  #8: ImageImporter ....................   Passed    0.01 sec
      Start  9: ride_ratings
 9/15 Test  #9: ride_ratings .....................   Passed    0.27 sec
      Start 10: multilaunch
10/15 Test #10: multilaunch ......................   Passed    0.85 sec
      Start 11: tile_elements
11/15 Test #11: tile_elements ....................   Passed    0.18 sec
      Start 12: replay_tests
12/15 Test #12: replay_tests .....................   Passed    8.87 sec
      Start 13: play_tests
13/15 Test #13: play_tests .......................   Passed    0.49 sec
      Start 14: pathfinding
14/15 Test #14: pathfinding ......................   Passed    0.43 sec
      Start 15: s6importexporttests
15/15 Test #15: s6importexporttests ..............   Passed    2.95 sec

100% tests passed, 0 tests failed out of 15

Total Test time (real) =  14.22 sec
```

janisozaur commented:
@ConiKost what's the status here? I can see some build issues in CI.

ConiKost commented:
@janisozaur I am unable to reproduce the errors of CI. I even setup a local i686 system, OpenRCT2 with my changes works in terms of gtest and running test.

>  /usr/bin/ld: i386:x86-64 architecture of input file `googletest-prefix/src/googletest-build/lib/libgtest.a(gtest-all.cc.o)' is incompatible with i386 output

That's the most curious here, like CI is mixing something arch specific? I don't have a single clou what to do here.

My local build is i386:
```
objdump -p googletest-prefix/src/googletest-build/lib/libgtest.a
In archive googletest-prefix/src/googletest-build/lib/libgtest.a:

gtest-all.cc.o:     file format elf32-i386
```

duncanspumpkin commented:
I'd like to see this working I can't actually run any tests anymore through Visual Studio due to the the replay tests causing issues. I'm somewhat tempted to scrap the way they currently work (test running every file) and just hard coding which files to run. That would mean we dont need anything fancy.

Gymnasiast commented:
Since this does not work, the original OP has not been able to reproduce this and it has been a year, I’ll close this. It can, of course, always be opened again if someone has any idea on how the fix the issues.
